### PR TITLE
Use worst-case data-dependent flops for AutoAC

### LINKED ISSUE #146707: AutoAC can cause deadlocks with tensor parallelism and data-dependent flop formulas
### 🐛 Describe the bug

Some operators' flop formulas depend on the content of their inputs, such as [FlashAttention when using a NestedTensor](https://github.com/pytorch/pytorch/blob/5d7532140f377195a831942f294c1f07c589bd9c/torch/utils/flop_counter.py#L298) (a.k.a varying sequence lengths, document-causal, block-diagonal, ...). The partitioner's AutoAC mode (enabled with `torch._functorch.config.activation_memory_budget < 1`) makes use of these formulas to choose which operator to checkpoint-vs-recompute in the backward.

This is already somewhat problematic (as it will "freeze" the decisions made with the first invocation's data for all subsequent invocations), but it becomes worse when the graph contains collectives, such as when using tensor/sequence parallelism. In such cases the partitioner might also introduce/suppress _collective_ ops from the backward graph, but if ranks have different data this could lead to inconsistent decisions and thus a desynchronization (a deadlock).

This is not just theoretical: it has occurred to us in practice, and here is a standalone repro:
```py
import tempfile
from datetime import timedelta
from functools import partial

import torch
import torch.distributed._functional_collectives as funcol
import torch.utils.flop_counter
from torch._inductor.utils import run_and_get_code

TOTAL_SEQLEN = 8192
NUM_HEADS = 8
HEAD_DIM = 128
EMB_DIM = NUM_HEADS * HEAD_DIM
HIDDEN_DIM = EMB_DIM * 2


def attn(xq, xk, xv, seqstarts, max_seqlen):
    return torch.ops.aten._flash_attention_forward(
        xq.unflatten(-1, (-1, HEAD_DIM)),
        xk.unflatten(-1, (-1, HEAD_DIM)),
        xv.unflatten(-1, (-1, HEAD_DIM)),
        seqstarts,
        seqstarts,
        max_seqlen,
        max_seqlen,
        0.0,
        True,
        return_debug_mask=False,
        scale=HEAD_DIM ** -0.5,
        window_size_left=-1,
        window_size_right=-1,
        seqused_k=None,
        alibi_slopes=None,
    )[0].flatten(1, 2)


def scale(t: torch.Tensor, amax_t: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
    max_v = torch.finfo(torch.float8_e4m3fn).max
    scale_t = torch.clamp(amax_t.float(), min=1e-12) / max_v
    t_fp8 = (t / scale_t).to(torch.float8_e4m3fn)
    return t_fp8, scale_t


def fp8_matmul(
    first: torch.Tensor,
    amax_first: torch.Tensor,
    second_t: torch.Tensor,
    amax_second_t: torch.Tensor,
    parallel: str,
) -> torch.Tensor:
    tp_group = torch.distributed.group.WORLD
    first_fp8, scale_first = scale(first, amax_first)
    second_t_fp8, scale_second_t = scale(second_t, amax_second_t)

    if parallel == "col":
        first_fp8 = funcol.all_gather_tensor(first_fp8, gather_dim=0, group=tp_group)
        scale_first = funcol.all_gather_tensor(scale_first, gather_dim=0, group=tp_group)

    res = torch._scaled_mm(
        first_fp8,
        second_t_fp8.t(),
        scale_a=scale_first,
        scale_b=scale_second_t.t(),
        out_dtype=torch.bfloat16,
    )

    if parallel == "row":
        res = funcol.reduce_scatter_tensor(res, "sum", scatter_dim=0, group=tp_group)

    return res


REVERSE = {"col": "row", "row": "col"}


@torch.compiler.allow_in_graph
class Fp8LinearFn(torch.autograd.Function):
    @staticmethod
    def forward(
        ctx: torch.autograd.function.FunctionCtx,
        a: torch.Tensor,
        b_t: torch.Tensor,
        parallel: str,
    ) -> torch.Tensor:
        amax_a = a.abs().amax(dim=-1, keepdim=True)
        amax_b_t = b_t.abs().amax(dim=-1, keepdim=True)
        out = fp8_matmul(a, amax_a, b_t, amax_b_t, parallel)
        ctx.save_for_backward(a, b_t, amax_b_t)
        ctx.parallel = parallel
        return out

    @staticmethod
    def backward(
        ctx: torch.autograd.function.FunctionCtx, grad_out: torch.Tensor
    ) -> tuple[torch.Tensor, torch.Tensor, None]:
        a: torch.Tensor
        b_t: torch.Tensor
        amax_b_t: torch.Tensor
        a, b_t, amax_b_t = ctx.saved_tensors
        parallel = REVERSE[ctx.parallel]

        # Workaround for https://github.com/pytorch/pytorch/issues/141881.
        b_t = b_t + grad_out[0, :, None]

        b = b_t.t().contiguous()
        amax_grad_out = grad_out.abs().amax(dim=-1, keepdim=True)
        amax_b = amax_b_t.t().amax(dim=-1, keepdim=True)
        amax_b = amax_b.repeat_interleave(
            b.shape[0] // amax_b.shape[0], dim=0, output_size=b.shape[0]
        )
        grad_a = fp8_matmul(
            grad_out, amax_grad_out, b, amax_b, parallel
        )

        tp_group = torch.distributed.group.WORLD
        if parallel == "col":
            grad_out = funcol.all_gather_tensor(grad_out, gather_dim=0, group=tp_group)
        if parallel == "row":
            a = funcol.all_gather_tensor(a, gather_dim=0, group=tp_group)
        grad_b = grad_out.t() @ a

        return grad_a, grad_b, None


@torch.compile(fullgraph=True)
def layer(x, wq, wk, wv, wo, seqstarts, max_seqlen):
    # y = funcol.all_gather_tensor_autograd(x, gather_dim=0, group=torch.distributed.group.WORLD)
    y = Fp8LinearFn.apply(
        attn(
            Fp8LinearFn.apply(x, wq, "col"),
            Fp8LinearFn.apply(x, wk, "col"),
            Fp8LinearFn.apply(x, wv, "col"),
            seqstarts,
            max_seqlen
        ),
        wo,
        "row"
    )
    # y = funcol.reduce_scatter_tensor_autograd(y, "sum", scatter_dim=0, group=torch.distributed.group.WORLD)
    return x + y


def run(rank: int, world_size: int, rdv_dir: str) -> None:
    torch.manual_seed(0)
    torch.cuda.set_device(rank)
    torch.distributed.init_process_group(
        backend="nccl",
        rank=rank,
        world_size=world_size,
        init_method=f"file://{rdv_dir}/rdv",
        timeout=timedelta(seconds=10),
    )

    torch._functorch.config.activation_memory_budget = 0.8005
    torch._dynamo.reset_code_caches()

    if rank == 0:
        # Least flops
        seqlens = [TOTAL_SEQLEN // 2, TOTAL_SEQLEN // 2]
    else:
        # Most flops
        seqlens = [TOTAL_SEQLEN - 128, 128]
    assert sum(seqlens) == TOTAL_SEQLEN
    max_seqlen = max(seqlens)
    seqlens = torch.tensor(seqlens)
    seqstarts = (seqlens.cumsum(dim=0) - seqlens).to(torch.int32).cuda()

    wq = torch.randn((EMB_DIM // world_size, EMB_DIM), dtype=torch.bfloat16, device="cuda")
    wk = torch.randn((EMB_DIM // world_size, EMB_DIM), dtype=torch.bfloat16, device="cuda")
    wv = torch.randn((EMB_DIM // world_size, EMB_DIM), dtype=torch.bfloat16, device="cuda")
    wo = torch.randn((EMB_DIM, EMB_DIM // world_size), dtype=torch.bfloat16, device="cuda")

    in_ = torch.randn((TOTAL_SEQLEN // world_size, EMB_DIM), dtype=torch.bfloat16, device="cuda")
    grad_out = torch.randn((TOTAL_SEQLEN // world_size, EMB_DIM), dtype=torch.bfloat16, device="cuda")

    (
        (out, (grad_in, grad_wq, grad_wk, grad_wv, grad_wo)),
        (fwd_code, bwd_code),
    ) = run_and_get_code(
        torch.autograd.functional.vjp,
        partial(layer, seqstarts=seqstarts, max_seqlen=max_seqlen),
        (in_, wq, wk, wv, wo),
        grad_out,
    )

    print(f"{rank=} {bwd_code.count("torch.ops._c10d_functional.all_gather_into_tensor.default")}")

    torch.distributed.destroy_process_group()


def main() -> None:
    world_size = 2
    with tempfile.TemporaryDirectory() as rdv_dir:
        torch.multiprocessing.spawn(
            run,
            args=(world_size, rdv_dir),
            nprocs=world_size,
        )


if __name__ == "__main__":
    main()
```

I believe that in a distributed setting AutoAC should use the "worst-case" flops in order to do its decisions, because:
- it will avoid deadlocks
- it will avoid the graph depending on the content of the data on the first invocation (flakiness, unpredictability)
- when using more and more GPUs, it will become more likely for one of them to hit the worst case (or very close to it), but since all GPUs go in lockstep they will anyways all pay that price, hence they should use that when planning

### Error logs

_No response_

### Versions

PyTorch nightly from PyPI, version `2.7.0.dev20250120+cu126`

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185820

AutoAC's memory-budget partitioner estimates recomputation cost with
FlopCounterMode. For varlen flash/efficient attention, the flop formulas can
read sequence offsets and max sequence lengths, so two distributed ranks with
the same packed tensor shape but different rank-local sequence layouts can get
different flop estimates. Those divergent estimates can produce different
save-vs-recompute decisions, which is especially dangerous when the joint graph
contains collectives.

Add an explicit data_dependent_flop_mode to FlopCounterMode and use
worst_case mode from the AutoAC runtime estimator. In this mode, varlen
attention formulas ignore rank-local offsets and data-derived max sequence
lengths, and instead estimate a shape-only worst case from the packed query,
key, and value lengths. The default flop counter behavior remains unchanged.

I considered changing the varlen attention formulas globally, but that would
remove useful actual flop accounting for users of FlopCounterMode. Scoping the
new mode to AutoAC keeps public behavior stable while making partition choices
independent of rank-local data.

Fixes #146707
Generated by my agent

Benchmark Results:
On a CUDA micro-benchmark running 100 compile-time flop-counter estimates for
_flash_attention_forward with packed shape [64, 2, 16] and seqlens [60, 4]:
- before-style actual mode: 0.021903s total, 462848 flops
- new worst_case mode: 0.009621s total, 524288 flops

Test Plan:
- python -m py_compile torch/utils/flop_counter.py torch/_functorch/partitioners.py test/functorch/test_ac.py test/test_flop_counter.py
- python test/functorch/test_ac.py -k test_flash_attention_autoac_uses_worst_case_flops
- python test/test_flop_counter.py -k test_nested_attention_worst_case_flop_mode (blocked during collection by local torchvision mismatch: operator torchvision::nms does not exist)
- python snippet masking torchvision as unavailable and running test/test_flop_counter.py test_nested_attention_worst_case_flop_mode
- python direct FlopCounterMode smoke test for actual vs worst_case varlen flash attention flops
- python CUDA micro-benchmark comparing actual vs worst_case FlopCounterMode estimator cost
- lintrunner -a torch/utils/flop_counter.py torch/_functorch/partitioners.py test/functorch/test_ac.py test/test_flop_counter.py

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this onto current main and addressed Claude's ContextVar cleanup suggestion by resetting the data-dependent flop mode token from a finally block in FlopCounterMode.__exit__. Focused tests and lintrunner passed locally.

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -3080,7 +3080,9 @@ def materialize_arg(x: Any) -> Any:
         from torch.utils.flop_counter import FlopCounterMode
 
         args, kwargs = pytree.tree_map(materialize_arg, (node.args, node.kwargs))
-        with FlopCounterMode(display=False) as mode:
+        with FlopCounterMode(
+            display=False, data_dependent_flop_mode="worst_case"
```
Comment (lw) [inline torch/_functorch/partitioners.py:3103]: Explain the context for this new flag? Or link to this PR or to the issue?

**Review by lw — APPROVED** [review]
Nice!

**Review by eellison — COMMENTED** [review]
I think this is an outdated fix. We should be doing a compiler collective to align on dynamic shape hints to fix this across ranks.

cc @ezyang @IvanKobzarev

File: torch/_functorch/partitioners.py (hunk 1 above)
Comment (jansel) [reply torch/_functorch/partitioners.py:3103]: My agent says I added a comment at the AutoAC flop-estimator callsite explaining that this mode keeps data-dependent flop estimates rank-invariant when AutoAC may add or remove collectives, and linked it to #146707.

**Comment by jansel** [comment]
My agent says I agree that sharing dynamic-shape hints through compiler collectives is the broader direction. I do not think that makes this scoped fix obsolete: today AutoAC uses the flop estimator here, and this issue comes from varlen attention flop formulas consuming rank-local data-derived values. This patch keeps that AutoAC decision rank-invariant for the concrete failure mode while leaving normal FlopCounter behavior unchanged.

**Comment by jansel** [comment]
@claude explain those CI failures, are they caused by this PR?
